server/src/main/java/org/elasticsearch/common/LocalTimeOffset.java

```diff
@@ -0,0 +1,679 @@
+     * offset for some local time that they have rounded down. In that case use
+     * {@link #localToSensibleUtc(long)} or {@link #localToUtc(long, Strategy)}.
+     */
+    public final long localToUtcInThisOffset(long localMillis) {
+        return localMillis - millis;
+    }
+
+    /**
+     * Convert a local time that occurs during this offset or a previous
+     * offset to utc, providing a strategy for how to resolve "funny" cases.
+     * You can use this if you've converted from utc to local, rounded down,
+     * and then want to convert back to utc and you need fine control over
+     * how to handle the "funny" edges.
+     * <p>
+     * This will not help you if you must convert a local time that you've
+     * rounded <strong>up</strong>. For that you are on your own. May God
+     * have mercy on your soul.
+     */
+    public abstract long localToUtc(long localMillis, Strategy strat);
+    public interface Strategy {
+        /**
+         * Handle a local time that never actually happened because a "gap"
+         * jumped over it. This happens in many time zones when folks wind
+         * their clocks forwards in the spring.
+         *
+         * @return the time in utc representing the local time
+         */
+        long inGap(long localMillis, Gap gap);
+        /**
+         * Handle a local time that happened before the start of a gap.
+         *
+         * @return the time in utc representing the local time
+         */
+        long beforeGap(long localMillis, Gap gap);
+        /**
+         * Handle a local time that happened twice because an "overlap"
+         * jumped behind it. This happens in many time zones when folks wind
+         * their clocks back in the fall.
+         *
+         * @return the time in utc representing the local time
+         */
+        long inOverlap(long localMillis, Overlap overlap);
+        /**
+         * Handle a local time that happened before the start of an overlap.
+         *
+         * @return the time in utc representing the local time
+         */
+        long beforeOverlap(long localMillis, Overlap overlap);
+    }
+
+    /**
+     * Map a local time that occurs during this offset or a previous offset
+     * to a utc time . You can use this if you've converted from utc to local,
+     * rounded down, and then want to convert back to utc but you don't really
+     * care about how to resolve "funny" edge cases. This resolves them
+     * <em>fairly</em> sanely.
+     * <p>
+     * If a local time occurred twice then returns the time in utc that the
+     * clocks first showed that time. If a local time never occurred then
+     * returns the utc time where the clock jumped over that time. If a local
+     * time occurs before a gap or overlap then uses the offset to convert it.
+     * This is all sensible but it may not be right for you.
+     */
+    public final long localToSensibleUtc(long localMillis) {
```

Comment:
"sensible" feels like the wrong word for timezone code. Suggest inlining this, it's only called once, and moving `SENSIBLE_STRAT` to a field of `ToMidnightRounding`, maybe renaming to `MIDNIGHT_STRATEGY` or similar.

Reply:
I liked having it on `LocalOffset` so I could test it without going through everything in `Rounding` and I think it'll take me a day or so to come around to removing it. Let me think on it!

If I don't end up moving it `MIDNIGHT` doesn't feel like the right name for it because it produces values that make *some* sense outside of the context of a local midnight. But, yeah, it probably does make sense to move it. I just have to admit that to myself.

Reply:
I've inlined the method into the rounding.